### PR TITLE
Html paragraph tag parse

### PR DESCRIPTION
**Pull request by DanElliottPalmer**

### CODE REVIEW

File: bbcondeparser/html_tags.py [hunk 1]
```diff
@@ -358,108 +343,235 @@ class BaseHTMLRenderTreeParser(BaseTreeParser):
...
+
+            elif isinstance(self._node, NewlineText):
+                self.handle_newline()
+
+        if self.is_inside_paragraph and self._paragraph_tree:
+            self.close_paragraph()
+
+        _tree = self._tree
+        self.stack_pop()
+        self._node.tree = _tree
```
Comment (neoshrew) [inline bbcondeparser/html_tags.py]: Need to set `parent` on the children in the tree.

File: bbcondeparser/html_tags.py [hunk 2]
```diff
@@ -358,108 +343,235 @@ class BaseHTMLRenderTreeParser(BaseTreeParser):
...
+    def open_paragraph(self, include_node=True):
+        self._inside_paragraph = True
+        if include_node:
+            self._paragraph_tree = [self._node]
+        else:
+            self._paragraph_tree = []
+
+    def close_paragraph(self):
+        # clone paragraph scope?
+        paragraph_node = ParagraphTag({}, self._paragraph_tree, '', '')
```
Comment (neoshrew) [inline bbcondeparser/html_tags.py]: Could we add a comment here to remind the reader that the `ParagraphTag` is going to set itself as the `parent` on the children in `self._paragraph_tree`?

File: bbcondeparser/html_tags.py [hunk 3]
```diff
@@ -358,108 +343,235 @@ class BaseHTMLRenderTreeParser(BaseTreeParser):
...
+
+    def open_paragraph(self, include_node=True):
+        self._inside_paragraph = True
+        if include_node:
+            self._paragraph_tree = [self._node]
+        else:
+            self._paragraph_tree = []
+
+    def close_paragraph(self):
+        # clone paragraph scope?
```
Comment (neoshrew) [inline bbcondeparser/html_tags.py]: What does this mean?

File: bbcondeparser/html_tags.py [hunk 4]
```diff
@@ -118,24 +121,12 @@ def validate_tag_cls(tag_cls):
         Raises:
             RuntimeError: if some values are set and shouldnt
         """
-        ctx_default = tag_cls.context_default
-        ctx_override = tag_cls.context_override
 
-        newline_behaviour = ctx_default.get('newline_behaviour')
-        if newline_behaviour not in NEWLINE_BEHAVIOURS:
+        if tag_cls.newline_behaviour not in NEWLINE_BEHAVIOURS:
             raise RuntimeError('newline_behaviour must be one of {}'.format(
```
Comment (neoshrew) [inline bbcondeparser/html_tags.py]: This should be a `ValueError`

This is my fault; I've used `RuntimeError` in other places and it's really not appropriate. Need to go back over the occurances of `RuntimeError` and switch them for more appropriate errors (probably `ValueError` or `TypeError`).

File: bbcondeparser/html_tags.py (hunk 3 above)
Comment (DanElliottPalmer) [reply bbcondeparser/html_tags.py]: leftover message from when i was working on it. i initially wasnt sure if i needed to create a copy of the paragraph tree before adding to the paragraph.

File: bbcondeparser/html_tags.py (hunk 1 above)
Comment (DanElliottPalmer) [reply bbcondeparser/html_tags.py]: ive done this but GH isnt picking it up 8a04b52b6fb7896a15f81cc6a35b2d473fb40f9b

File: bbcondeparser/html_tags.py [hunk 5]
```diff
@@ -138,27 +142,71 @@ def validate_tag_cls(tag_cls):
 
 
 def get_newline_behaviour(current_value, new_value):
+    """Return the newline behaviour based on current and new values.
+
+    NEWLINE_BEHAVIOURS is used to store what behaviours take privaledge.
```
Comment (neoshrew) [inline bbcondeparser/html_tags.py]: *privilege

**Pull request merged**